w3af/plugins/infrastructure/webapp_detection.py

```diff
@@ -0,0 +1,139 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+"""
+from wad.detection import Detector, TIMEOUT
+from wad.output import HumanReadableOutput
+
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.kb.knowledge_base as kb
+
+from w3af.core.controllers.plugins.infrastructure_plugin import InfrastructurePlugin
+from w3af.core.controllers.exceptions import RunOnce
+from w3af.core.controllers.misc.decorators import runonce
+from w3af.core.data.kb.info import Info
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options.option_list import OptionList
+
+
+class w3afDetector(Detector):
+    """
+    This class overrides some getter functions in original Detector class
+    in order to use w3af's ExtendedUrllib.
+    """
+    def __init__(self, uri_opener):
+        super(w3afDetector, self).__init__()
+        self._uri_opener = uri_opener
+
+    def get_page(self, url, timeout=TIMEOUT):
+        return self._uri_opener.GET(url, cache=True, timeout=timeout)
+
+    def get_new_url(self, page):
+        return page.get_url().url_string
+
+    def get_content(self, page, url):
+        return page.get_body()
+
+
+class webapp_detection(InfrastructurePlugin):
+    """
+    Identify technologies used by website based on HTTP response and HTML content
+
+    This plugin is a wrapper of CERN's CERT team's Web application detection (WAD) tool.
+
+    :author: Piotr Lizonczyk (piotr.lizonczyk@gmail.com)
+    """
+    def __init__(self):
+        InfrastructurePlugin.__init__(self)
+
+        # User configured
+        self._timeout = 5
+
+    @runonce(exc_class=RunOnce)
+    def discover(self, fuzzable_request):
+        """
+        It calls the "main" from WAD and writes the results to the kb.
+
+        :param fuzzable_request: A fuzzable_request instance that contains
+                                    (among other things) the URL to test.
+        """
+        msg = 'Web application detection (WAD) plugin is starting. ' \
+              'Original WAD authors: CERN-CERT; http://cern.ch/security'
+        om.out.information(msg)
+
+        self._main(fuzzable_request.get_url())
+
+    def _main(self, url):
+        """
+        Based on WAD's main executable
+        """
+
+        detector = w3afDetector(uri_opener=self._uri_opener)
+        results = detector.detect(url, timeout=self._timeout)
+
+        self._report(results)
+
+    def _report(self, results):
+        """
+        Displays detailed report information to the user and save the data to
+        the kb.
+
+        :return: None.
+        """
+        if results:
+            i = Info('Web applications detection report',
+                     HumanReadableOutput().retrieve(results), 1,
```

Comment:
Please replace the 1 with the real HTTP request ID used to identify this information